x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/TransportGetDeploymentStatsAction.java

```diff
@@ -0,0 +1,142 @@
+import org.elasticsearch.xpack.core.action.util.ExpandedIdsMatcher;
+import org.elasticsearch.xpack.core.ml.MlTasks;
+import org.elasticsearch.xpack.core.ml.action.GetDeploymentStatsAction;
+import org.elasticsearch.xpack.core.ml.utils.ExceptionsHelper;
+import org.elasticsearch.xpack.ml.inference.deployment.DeploymentManager;
+import org.elasticsearch.xpack.ml.inference.deployment.ModelStats;
+import org.elasticsearch.xpack.ml.inference.deployment.TrainedModelDeploymentTask;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+public class TransportGetDeploymentStatsAction extends TransportTasksAction<TrainedModelDeploymentTask,
+    GetDeploymentStatsAction.Request, GetDeploymentStatsAction.Response, GetDeploymentStatsAction.Response.DeploymentStats> {
+
+    private final DeploymentManager deploymentManager;
+
+    @Inject
+    public TransportGetDeploymentStatsAction(TransportService transportService, ActionFilters actionFilters, ClusterService clusterService,
+                                             DeploymentManager deploymentManager) {
+        super(GetDeploymentStatsAction.NAME, clusterService, transportService, actionFilters, GetDeploymentStatsAction.Request::new,
+            GetDeploymentStatsAction.Response::new, GetDeploymentStatsAction.Response.DeploymentStats::new, ThreadPool.Names.MANAGEMENT);
+
+        this.deploymentManager = deploymentManager;
+    }
+
+    @Override
+    protected GetDeploymentStatsAction.Response newResponse(GetDeploymentStatsAction.Request request,
+                                                            List<GetDeploymentStatsAction.Response.DeploymentStats> tasks,
+                                                            List<TaskOperationFailure> taskOperationFailures,
+                                                            List<FailedNodeException> failedNodeExceptions) {
+        tasks.sort(Comparator.comparing(GetDeploymentStatsAction.Response.DeploymentStats::getModelId));
+        return new GetDeploymentStatsAction.Response(taskOperationFailures, failedNodeExceptions, tasks, tasks.size());
+    }
+
+    @Override
+    protected void doExecute(Task task, GetDeploymentStatsAction.Request request,
+                             ActionListener<GetDeploymentStatsAction.Response> listener) {
+
+        ClusterState clusterState = clusterService.state();
+        PersistentTasksCustomMetadata tasks = clusterState.getMetadata().custom(PersistentTasksCustomMetadata.TYPE);
+
+        String[] tokenizedRequestIds = Strings.tokenizeToStringArray(request.getDeploymentId(), ",");
+
+        Collection<PersistentTasksCustomMetadata.PersistentTask<?>> deploymentTasks = MlTasks.trainedModelDeploymentTasks(tasks);
+
+        if (Strings.isAllOrWildcard(request.getDeploymentId())) {
```

Comment:
Do we need to do all this or could we not solve this doing it similarly to `TransportStopDataFrameAnalyticsAction.matchAllStartedIds` method?

Reply:
looking at this again I found a way to simplify the code.